contracts/src/PriceFeeds/CompositePriceFeed.sol

```diff
@@ -4,73 +4,72 @@ pragma solidity 0.8.18;
-import "../Interfaces/ICompositePriceFeed.sol";
 
 // import "forge-std/console2.sol";
 
-// Composite PriceFeed: outputs an LST-USD price derived from two external price Oracles: LST-ETH, and ETH-USD.
-// Used where the LST token is non-rebasing (as per rETH, osETH, ETHx, etc).
-contract CompositePriceFeed is MainnetPriceFeedBase, ICompositePriceFeed {
-    Oracle public lstEthOracle;
-    Oracle public ethUsdOracle;
-
+// The CompositePriceFeed is used for feeds that incorporate both a market price oracle (e.g. STETH-USD, or RETH-ETH) 
+// and an LST canonical rate (e.g. WSTETH:STETH, or RETH:ETH).
+contract CompositePriceFeed is MainnetPriceFeedBase {
     address public rateProviderAddress;
 
     constructor(
         address _owner,
         address _ethUsdOracleAddress,
-        address _lstEthOracleAddress,
         address _rateProviderAddress,
-        uint256 _ethUsdStalenessThreshold,
-        uint256 _lstEthStalenessThreshold
-    ) MainnetPriceFeedBase(_owner) {
-        // Store ETH-USD oracle
-        ethUsdOracle.aggregator = AggregatorV3Interface(_ethUsdOracleAddress);
-        ethUsdOracle.stalenessThreshold = _ethUsdStalenessThreshold;
-        ethUsdOracle.decimals = ethUsdOracle.aggregator.decimals();
-        assert(ethUsdOracle.decimals == 8);
-
-        // Store LST-ETH oracle
-        lstEthOracle.aggregator = AggregatorV3Interface(_lstEthOracleAddress);
-        lstEthOracle.stalenessThreshold = _lstEthStalenessThreshold;
-        lstEthOracle.decimals = lstEthOracle.aggregator.decimals();
-
+        uint256 _ethUsdStalenessThreshold
+    ) MainnetPriceFeedBase(
+        _owner,
+        _ethUsdOracleAddress,
+        _ethUsdStalenessThreshold
+    ) {
         // Store rate provider
         rateProviderAddress = _rateProviderAddress;
-
-        _fetchPrice();
-
-        // Check an oracle didn't already fail
-        assert(priceFeedDisabled == false);
+    }
+    
+    function fetchPrice() public returns (uint256, bool) {
+        // If branch is live and the primary oracle setup has been working, try to use it 
+        if (priceSource == PriceSource.primary) {return _fetchPrice();}
+
+        // If branch is already shut down and using ETH-USD * canonical_rate, try to use that
+        if (priceSource == PriceSource.ETHUSDxCanonical) {
+                (uint256 ethUsdPrice, bool ethUsdOracleDown) = _getOracleAnswer(ethUsdOracle);
+                //... but if the ETH-USD oracle *also* fails here, use the lastGoodPrice
+                if(ethUsdOracleDown) {
+                    // No need to shut down, since branch already is shut down
+                    return (lastGoodPrice, false);
+                } else {
+                    return (_fetchPriceETHUSDxCanonical(ethUsdPrice), false);
+                }
+            }
+
+        // Otherwise if branch is shut down and already using the lastGoodPrice, continue with it
+        if (priceSource == PriceSource.lastGoodPrice) {return (lastGoodPrice, false);}
```

Comment:
Maybe we can get rid of the `if`, as there’s no other possibility, and note it in a comment.
It doesn’t  matter too much, but right now it feels like the execution flow could not enter in any of the `if` clauses and return `(0, false)` (which is not the case, fortunately).

Reply:
This is not done yet, right?